<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Анализ-базы-данных-сервиса-для-чтения-книг-по-подписке" data-toc-modified-id="Анализ-базы-данных-сервиса-для-чтения-книг-по-подписке-1">Анализ базы данных сервиса для чтения книг по подписке</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.0.1">Описание данных</a></span></li><li><span><a href="#Задачи:" data-toc-modified-id="Задачи:-1.0.2">Задачи:</a></span></li><li><span><a href="#Сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Сколько-книг-вышло-после-1-января-2000-года-1.0.3">Сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку-1.0.4">Для каждой книги посчитать количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определение-издательства,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-для-исключения-из-анализа-брошюры" data-toc-modified-id="Определение-издательства,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-для-исключения-из-анализа-брошюры-1.0.5">Определение издательства, которое выпустило наибольшее число книг толще 50 страниц для исключения из анализа брошюры</a></span></li><li><span><a href="#Определение-автора-с-самой-высокой-средней-оценкой-книг,-учитываются-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определение-автора-с-самой-высокой-средней-оценкой-книг,-учитываются-только-книги-с-50-и-более-оценками-1.0.6">Определение автора с самой высокой средней оценкой книг, учитываются только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитать-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок" data-toc-modified-id="Посчитать-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок-1.0.7">Посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.1">Вывод</a></span></li></ul></li></ul></div>

# Анализ базы данных сервиса для чтения книг по подписке

В базе данных содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Наша задача проанализировать данные для формулирования ценностное предложение для нового продукта. 

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Задачи: 
- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц для исключения из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг, учитываются только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок.


In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


In [3]:
def sql_query(df):
    query = '''SELECT * FROM {} LIMIT 5'''.format(df)
    con=engine.connect()
    display(pd.io.sql.read_sql(sql=text(query), con = con))
    

In [4]:
for df in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    sql_query(df)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

### Сколько книг вышло после 1 января 2000 года


In [5]:
query = '''
SELECT COUNT(book_id) 
FROM books 
WHERE publication_date > '2000-01-01'
    '''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

После 01.01.2000 вышло 819 книг

### Для каждой книги посчитать количество обзоров и среднюю оценку



In [6]:
query = '''
SELECT     b.book_id,
    title,
    COUNT(DISTINCT  re.review_id) count_rewiew,
    ROUND(AVG(ra.rating),1) avg_score
FROM books AS b 
LEFT JOIN ratings AS ra ON b.book_id=ra.book_id
LEFT JOIN reviews AS re ON b.book_id=re.book_id
GROUP BY b.book_id
ORDER BY count_rewiew DESC
    '''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

book_id                                              title  count_rewiew  \
0        948                            Twilight (Twilight  #1)             7   
1        963                                Water for Elephants             6   
2        734                                   The Glass Castle             6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...             6   
4        695  The Curious Incident of the Dog in the Night-Time             6   
..       ...                                                ...           ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
996      808                            The Natural Way to Draw             0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites             0   
998      221                          Essential Tales and Poems             0   
999      191  Disney's Beauty and the Beast (A Little Golden...             0   

     avg_score  
0          3.7  
1          4.0  
2          4.2  
3          4.4  
4          4.1  
..         ...  
995        3.7  
996        3.0  
997        5.0  
998        4.0  
999        4.0  

[1000 rows x 4 columns]

В базе содержится 1000 книг, для большинства из них мы получили количество обзоров  и средний рейтинг. Максимальное число обзоров 7 на книгу, не на все книги есть обзор.

### Определение издательства, которое выпустило наибольшее число книг толще 50 страниц для исключения из анализа брошюры


In [7]:
query = '''
SELECT publisher,
    COUNT (book_id) count_book
FROM books AS b 
JOIN publishers AS p ON b.publisher_id=p.publisher_id
WHERE num_pages > 50
GROUP BY publisher
ORDER BY count_book DESC
LIMIT 1
    '''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

publisher  count_book
0  Penguin Books          42

Больше всего книг (42) выпустило издательство Penguin Books.

### Определение автора с самой высокой средней оценкой книг, учитываются только книги с 50 и более оценками


In [8]:
query = '''
SELECT author,
    ROUND(AVG(rating),1) avg_raiting
FROM books AS b
JOIN authors AS a ON b.author_id=a.author_id
JOIN ratings AS r ON b.book_id=r.book_id
GROUP BY author
HAVING COUNT(rating_id)>50
ORDER BY avg_raiting DESC
LIMIT 1
    '''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

author  avg_raiting
0  J.K. Rowling/Mary GrandPré          4.3

Автор с самым большим рейтингом и количеством оценок больше 50 это J.K. Rowling/Mary GrandPré.

### Посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [9]:
query = '''
SELECT AVG(count_rewiew)
FROM (SELECT username,
        COUNT(review_id) count_rewiew
    FROM reviews
    WHERE username IN  (SELECT username
                        FROM ratings
                        GROUP BY username
                        HAVING COUNT(rating) > 48)
    GROUP BY username) AS one
    '''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)


avg
0  24.0

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок равно 24.

## Вывод
Наша база данных состоит из 5 баз с информацией о книгах, авторах, издательствах, а также об отзывах и рейтингах читателей. 

В базе есть информация о 1000 книг. Популярность книг мы можем оценить по рейтингам читателей, а на некоторые книги у нас даже есть развернутые отзывы (целых 7 отзывов на самую популярную книгу Сумерки). Больше всего книг выпустило издательсво Penguin Books - 42 книги, при расчетах мы исключили брошюры (книги, где меньше 50 страниц). Самый популярный автор это Дж.К. Роулинг. Также мы можем оценить, что те, кто активно оценивает книги, те еще и активно пишут отзывы - в среднем 24 отзыва. Также судя по неймингу, можно сделать вывод, что сервис рассчитаан на англоязычную публику.